### PR TITLE
DFT and Scaled Endpoint Design document

### PR DESCRIPTION
**Pull request by phudtran**

### CODE REVIEW

File: docs/design/dft.md [hunk 1]
```diff
@@ -0,0 +1,1007 @@
...
+
+-   Overlay protocol integrated within the data-plane to provide
+  consistent connection state
+
+-   Built-in multi-tenancy that allows infrastructure sharing of network
+  services and middleboxes without the need to isolate middleboxes in
+  virtual machines or containers, hence improves cost
+
+-   Native data-plane failure detection that minimizes the time to
+  recover a connection by means of gossip proto- cols
```
Comment (deepak-vij) [inline docs/design/dft.md]: proto- cols -? protocols

File: docs/design/dft.md [hunk 2]
```diff
@@ -0,0 +1,1007 @@
...
+
+# Software Organization
+
+We will manages two separate pools of nodes. The
```
Comment (deepak-vij) [inline docs/design/dft.md]: manages -> manage

File: docs/design/dft.md [hunk 3]
```diff
@@ -0,0 +1,1007 @@
...
+first pool consists of data-plane nodes (D-Nodes) that are responsible
+for all packet processing. The other pool consists of management-plane
+nodes (M-Nodes) that run a set of Kubernetes custom resource operators
```
Comment (deepak-vij) [inline docs/design/dft.md:28]: that run -> that runs

File: docs/design/dft.md [hunk 4]
```diff
@@ -0,0 +1,1007 @@
...
+and Kubernetes services . The operators implement the management- plane logic,
```
Comment (deepak-vij) [inline docs/design/dft.md]: management- plane -> management-plane

File: docs/design/dft.md [hunk 5]
```diff
@@ -0,0 +1,1007 @@
...
+and the other services cooperate with the operators to manage the
+D-Nodes and the middleboxes as we will detail in the Management Plane
```
Comment (deepak-vij) [inline docs/design/dft.md:31]: as we will detail -> as explained in

File: docs/design/dft.md [hunk 6]
```diff
@@ -0,0 +1,1007 @@
...
+Section. The D,M-Nodes may join and leave the cluster dynamically,
+and the D-Nodes are not necessarily Kubernetes worker nodes and can
+be external to a Kubernetes cluster.
+
+The management-plane exposes network services as endpoints of a single
+IP address (VIP), each implements a chain of one or more middleboxes.
+We refer to such network service as (Scaled Endpoint),as an endpoint
```
Comment (deepak-vij) [inline docs/design/dft.md]: ,as -> , as

File: docs/design/dft.md [hunk 7]
```diff
@@ -0,0 +1,1007 @@
...
+scale-out and in across multiple D-Nodes according to its bandwidth
+demand. The following is a non-exhaustive list of poten- tial Scaled
```
Comment (deepak-vij) [inline docs/design/dft.md]: demand -> demands

Comment (deepak-vij) [inline docs/design/dft.md]: poten- tial -> potential

File: docs/design/dft.md [hunk 8]
```diff
@@ -0,0 +1,1007 @@
...
+Endpoint types: Bouncer, Divider, Load-balancer, Nat, DPI, QoS,
+Tunnel, Firewall, Protocol Translators IPv6/IPv4, Access Control List
+(ACL) or a combination of these. A typical example from production
+systems is a Scaled Endpoint that implements a secure load-balancer
+and consists of two middleboxes: ACL Loadlancer). Developers shall
```
Comment (deepak-vij) [inline docs/design/dft.md]: ACL Loadlancer) -> ACL & Load-balancer

File: docs/design/dft.md [hunk 9]
```diff
@@ -0,0 +1,1007 @@
...
+be able to specify and verify different Scaled Endpoints
+functionalities through a specification inter- face.
```
Comment (deepak-vij) [inline docs/design/dft.md]: inter- face -> interface

File: docs/design/dft.md [hunk 10]
```diff
@@ -0,0 +1,1007 @@
...
+and the other services cooperate with the operators to manage the
+D-Nodes and the middleboxes as we will detail in the Management Plane
+Section. The D,M-Nodes may join and leave the cluster dynamically,
+and the D-Nodes are not necessarily Kubernetes worker nodes and can
+be external to a Kubernetes cluster.
+
+The management-plane exposes network services as endpoints of a single
+IP address (VIP), each implements a chain of one or more middleboxes.
+We refer to such network service as (Scaled Endpoint),as an endpoint
+scale-out and in across multiple D-Nodes according to its bandwidth
```
Comment (deepak-vij) [inline docs/design/dft.md:39]: scale-out and in -> scale-out/in

File: docs/design/dft.md (hunk 9 above)
Comment (deepak-vij) [inline docs/design/dft.md]: functionalities -> functionality

Comment (deepak-vij) [reply docs/design/dft.md]: @zasherif , I have started reviewing the design document. Found few minor grammatical type problems. I will keep reviewing it further and provide feedback accordingly.

Comment (zasherif) [reply docs/design/dft.md]: Sure. Thank you very much, Deepak!

File: docs/design/dft.md [hunk 11]
```diff
@@ -0,0 +1,1007 @@
...
+demand. The following is a non-exhaustive list of potential Scaled
+Endpoint types: Bouncer, Divider, Load-balancer, Nat, DPI, QoS,
+Tunnel, Firewall, Protocol Translators IPv6/IPv4, Access Control List
+(ACL) or a combination of these. A typical example from production
+systems is a Scaled Endpoint that implements a secure load-balancer
+and consists of two middleboxes: ACL Loadlancer). Developers shall
+be able to specify and verify different Scaled Endpoints
+functionalities through a specification interface.
```
Comment (deepak-vij) [inline docs/design/dft.md:47]: functionalities -> functionality

File: docs/design/dft.md [hunk 12]
```diff
@@ -0,0 +1,1007 @@
...
+
+# Dataplane Nodes
+
+D-Nodes (Figure [1] are Virtual Machines (VMs) or SmartNiCs
+that support turnkey Linux distributions or XDP offloading
+(e.g., Broadcom PS225 , Netronome Agilio. Each D-Node may
```
Comment (deepak-vij) [inline docs/design/dft.md]: (e.g., Broadcom PS225 , Netronome Agilio. -> (e.g., Broadcom PS225 , Netronome Agilio etc.).

File: docs/design/dft.md [hunk 13]
```diff
@@ -0,0 +1,1007 @@
...
+have multiple network interfaces. The management-plane abstracts
+network interfaces as droplets running on the D-Node. A droplet is
+a one-to-one mapping to a network interface. A droplet daemon
+(droplet-d) manages the dropletswhere it attaches an XDP program to
```
Comment (deepak-vij) [inline docs/design/dft.md]: dropletswhere it attaches an XDP program to -> droplets by attaching a XDP program to

File: docs/design/dft.md [hunk 14]
```diff
@@ -0,0 +1,1007 @@
...
+each interface, populates ebpf maps storing configuration, flow-caching,
+and metrics. The droplet-d also manages data plane userspace
+middleboxes, where packet processing may be offloaded to the userspace
+using AF XDP. The D-Nodesalso deploy a nomad driver and device plugins
```
Comment (deepak-vij) [inline docs/design/dft.md]: The D-Nodesalso  deploy -> The D-Nodes also deploys

Comment (deepak-vij) [inline docs/design/dft.md]: nomad driver and device plugins -> nomad driver and relevant device plugins

File: docs/design/dft.md [hunk 15]
```diff
@@ -0,0 +1,1007 @@
...
+that extends Nomad/Consul  for health checks, telemetry, placement, and
```
Comment (deepak-vij) [inline docs/design/dft.md]: that extends Nomad/Consul  for -> as part of Nomad/Consul to perform activities such as

File: docs/design/dft.md [hunk 16]
```diff
@@ -0,0 +1,1007 @@
...
+autoscaling. In the case of SmartNics, the D-Node's management plane
+components may be deployed to a common server -- installing multiple
+SmartNics -- which acts as aproxy to manage the dataplane through an RPC interface.
```
Comment (deepak-vij) [inline docs/design/dft.md]: aproxy -> a proxy

Comment (deepak-vij) [inline docs/design/dft.md]: an RPC -> a RPC

File: docs/design/dft.md [hunk 17]
```diff
@@ -0,0 +1,1007 @@
...
+
+
+<img src="png/dnode.png" width="300" height="400">
+
+Figure 1: Dataplane Node Worker Architecture
+
+# Droplet Organization
+
+Each droplet runs an XDP program that implements the data-plane
```
Comment (deepak-vij) [inline docs/design/dft.md:74]: an XDP -> a XDP

File: docs/design/dft.md [hunk 18]
```diff
@@ -0,0 +1,1007 @@
...
+services are the IPRewriter, IPClassifier, and IPFilter. We
+introduce two new M-XDP types that are specific to the cluster
+functionalities; Forwarder (FWD) and Flow Table Node (FTN)
+
+<img src="png/xdp.png" width="300" height="150">
+
+Figure 2: Droplet Architecture
+
+An M-XDP entirely runs in the kernel or offloaded to a SmartNIC. For
+some complex middleboxes , may send packets to userspace through AF
```
Comment (deepak-vij) [inline docs/design/dft.md]: complex middleboxes , -> complex middleboxes,

File: docs/design/dft.md [hunk 19]
```diff
@@ -0,0 +1,1007 @@
...
+XDP in case the D-Node is a Virtual Machine. The P-XDP and M-XDPs all
+share the same ebpf maps. The management plane provides packet
+classification rules by which a P-XDP determines which M-XDP shall
+handles the packet and tail-calls the corresponding ebpf program. A
```
Comment (deepak-vij) [inline docs/design/dft.md]: handles -> handle

Comment (deepak-vij) [inline docs/design/dft.md]: and tail-calls -> and subsequently tail-calls

File: docs/design/dft.md [hunk 20]
```diff
@@ -0,0 +1,1007 @@
...
+elements. The most important elements for most of the network
+services are the IPRewriter, IPClassifier, and IPFilter. We
+introduce two new M-XDP types that are specific to the cluster
+functionalities; Forwarder (FWD) and Flow Table Node (FTN)
+
+<img src="png/xdp.png" width="300" height="150">
+
+Figure 2: Droplet Architecture
+
+An M-XDP entirely runs in the kernel or offloaded to a SmartNIC. For
```
Comment (deepak-vij) [inline docs/design/dft.md:88]: An M-XDP -> A M-XDP

File: docs/design/dft.md [hunk 21]
```diff
@@ -0,0 +1,1007 @@
...
+
+-   **Metering maps (m-map)**:
+  Read/Write maps used by middleboxes to provide packet and health
+  metrics.
+
+# Management Plane Nodes
+
+Users, administrators, and management plane operators interact with a
+the cluster through the Kubernetes API server. Through
+Custom Resource Definition , we extend the API
```
Comment (deepak-vij) [inline docs/design/dft.md]: Resource Definition , -> Resource Definition,

File: docs/design/dft.md [hunk 22]
```diff
@@ -0,0 +1,1007 @@
...
+server functionalities to manage new types of objects.
+Each object has a custom resource operator that
+manages its life-cycle from creation, to provisioning, to deletion
+through various workflows. An administrator deploys an operator as a
+pod, that's integrated in the control-loop of Kubernetes.
+
+The management-plane employs a Nomad and Consul clusters to provide
```
Comment (deepak-vij) [inline docs/design/dft.md]: Consul clusters -> Consul cluster

File: docs/design/dft.md [hunk 23]
```diff
@@ -0,0 +1,1007 @@
...
+off-the-shelf services for large scale clusters of several thousands
```
Comment (deepak-vij) [inline docs/design/dft.md]: thousands -> thousands of

File: docs/design/dft.md [hunk 24]
```diff
@@ -0,0 +1,1007 @@
...
+D-Nodes. Similar to operators, an administrator shall deploy
+Nomad/Consul clusters as Kubernetes Pods. Nomad exposes
```
Comment (deepak-vij) [inline docs/design/dft.md:139]: clusters -> cluster

File: docs/design/dft.md [hunk 25]
```diff
@@ -0,0 +1,1007 @@
...
+ensures that we can deploy with legacy cloud
+networks as well as state-of-the art host networking approaches.
+
+<img src="png/selector.png" width="300" height="250">
+
+Figure 4: Scaled Endpoint
+
+The Scaled Endpoint operator keeps routers, or host switches updated
+with the set of FWDs of a Scaled Endpoint. For example, Mizar uses a
+simple tuple hashing in each Transit Agent - an XDP program that
```
Comment (deepak-vij) [inline docs/design/dft.md:219]: an XDP -> a XDP

File: docs/design/dft.md [hunk 26]
```diff
@@ -0,0 +1,1007 @@
...
+encapsulates VMs/Container egress traffic ­ -- to select a bouncer
+from a set of bouncer's FWDs. Mizar provides an RPC interface to
```
Comment (deepak-vij) [inline docs/design/dft.md:221]: an RPC -> a RPC

File: docs/design/dft.md [hunk 27]
```diff
@@ -0,0 +1,1007 @@
...
+Figure 10: Regular packet flow
+
+**Packet Flow:** Algorithm 1 and Algorithm
+2 detail the FWD and FTN behaviors for packet
+processing. Figure [10] describes a typical example of
+the regular packet flow and interaction between the FWDs and the DFT.
+The packet progresses from source to destination as follows:
+
+1. FWD receives a packet from source
+2. FWD lookup the packet's flow entry in its flow f-map (cached)
```
Comment (deepak-vij) [inline docs/design/dft.md]: FWD lookup the packet's flow entry in its flow f-map (cached) -> FWD does the packet's flow entry lookup in its flow f-map (cached)

File: docs/design/dft.md [hunk 28]
```diff
@@ -0,0 +1,1007 @@
...
+
+    2.  If no entry exists, fwd hashes the packet's tuple and selects
+        the corresponding chain from the Maglev table, add an NSH header
+        corresponding to the Scaled Endpoint of the packet (with a
+        Lookup Opcode), then sends the packet to the tail FTN of the
+        chain (FTN-T3 in Figure [10]. If the packet is not
+        encapsulated (external traffic), the FWD
+        encapsulates the packet.
+
+3.  When the tail FTN receives packet 2(b), it lookup the packet flow
```
Comment (deepak-vij) [inline docs/design/dft.md]: it lookup the packet flow -> it does a lookup of the packet flow

File: docs/design/dft.md [hunk 29]
```diff
@@ -0,0 +1,1006 @@
...
+be external to a Kubernetes cluster.
+
+The management-plane exposes network services as endpoints of a single
+IP address (VIP), each implements a chain of one or more middleboxes.
+We refer to such network service as (Scaled Endpoint), as an endpoint
+scale-out and in across multiple D-Nodes according to its bandwidth
+demands. The following is a non-exhaustive list of potential Scaled
+Endpoint types: Bouncer, Divider, Load-balancer, Nat, DPI, QoS,
+Tunnel, Firewall, Protocol Translators IPv6/IPv4, Access Control List
+(ACL) or a combination of these. A typical example from production
```
Comment (Eric-Yuan) [inline docs/design/dft.md:43]: If it is possible that we could provide a mechnism that let tenant users can choose their custom developed network service software as m-nodes? Or provide a open m-node framework, which tenant users can base on to develop their own m-nodes?

Comment (zasherif) [reply docs/design/dft.md:43]: I think this is a very good suggestion, we can certainly extend the m-node functionalities to allow tenants (in addition to admins) to define services themselves using basic building blocks (m-xdps) each doing a certain function. Currently, we intended to give this option to admins, but it's very possible to also extend it to users.

**Pull request closed without merging**